fix(validator): fix jsons validation after arrays

dario-piotrowicz opened the pull request:
fix the bug of `v.json()` calls made after a `v.array()` to be mistakenly considered as still being inside the array

fixes one of the issues mentioned in #677

____

This is the result of the test I've added before applying the change:
![before](https://user-images.githubusercontent.com/61631103/204113972-97527d99-f89b-4a7f-bc40-f1c42d1c031d.png)

(the rest of the checks passed already, but I've included them for completeness and to cover all cases for `v.json()` :slightly_smiling_face: )

yusukebe commented on src/validator/validator.test.ts:
```diff
@@ -730,6 +730,63 @@ describe('Nested objects', () => {
+
+  it('Should threat `v.json()` as a nested array type only inside an array', async () => {
+    const json = {
+      blog: {
```
If we add the `author.name` such as below:

```ts
      blog: {
        totalPosts: '3',
        posts: [
          {
            id: 123,
            title: 'JavaScript',
            author: {
              name: 'foo',
            },
          },
          {
            id: '456',
            title: 'Framework',
            author: {
              name: 'bar',
            },
          },
          {
            id: 789,
            title: 'Hono',
            author: {
              name: 'baz',
            },
          },
        ],
        highlightedPostId: '123',
      },
```

and write the validator below:

```ts
    const v = new Validator()
    const vObject = v.object('blog', (v) => ({
      posts: v.array('posts', (v) => ({
        author: v.object('author', (v) => ({
          name: v.json('name'),
        })),
      })),
    }))
```

then, we expect it to be valid, but it will no be valid. So, I think we have to remove this line:

https://github.com/honojs/hono/blob/8ed7763c38e153ee78860724582ffd0f0d96a845/src/validator/validator.ts#L111

dario-piotrowicz replied:
good point, I guess it was added there to reset the `isArray` field? that shouldn't be necessary anymore right?

yusukebe replied:
Yes, this line is not needed.

yusukebe replied:
The tests you write look good!

yusukebe commented:
Hi @dario-piotrowicz !

Thank you for the PR. I'm just trying to fix this issue. I've added the comment, please check it!

dario-piotrowicz commented:
> Hi @dario-piotrowicz !
> 
> Thank you for the PR. I'm just trying to fix this issue. I've added the comment, please check it!

Cool, good point thanks :slightly_smiling_face:
I've applied your suggested change and updated the unit test, all seems to be working fine :slightly_smiling_face::+1:

yusukebe commented on src/validator/validator.ts:
```diff
@@ -104,10 +104,10 @@ export class Validator {
     this.isArray = true
     const res = validator(this)
     const arr = new VArray(res, path)
+    this.isArray = false
```
I reconsidered and think it would be better to create a new instance of `Validator`.

```ts
  array = <T extends Schema>(path: string, validator: (v: Validator) => T): VArray<T> => {
    const newValidator = new Validator()
    newValidator.isArray = true
    const res = validator(newValidator)
    const arr = new VArray(res, path)
    return arr
  }
  object = <T extends Schema>(path: string, validator: (v: Validator) => T): VObject<T> => {
    const newValidator = new Validator()
    newValidator.isArray = this.isArray
    const res = validator(newValidator)
    const obj = new VObject(res, path)
    return obj
  }
```

In this case, a different object from `this` should be invoked in `validator`.

dario-piotrowicz replied:
yes I totally agree!

To be honest I thought of this but I wasn't sure there was a specific reason why the same instance of `Validator` was being used (I guess it is more efficient, although I can't imagine it making a noticeable impact on performances).

I'll make the change :slightly_smiling_face::+1:

dario-piotrowicz replied:
@yusukebe I've changed the implementation, I soft of followed your suggestion but changed things up a bit, I thinks (please let me know if I'm wrong) that `isArray` should not be accessible from the outside of the class (as the user can access that as well).
So I made the `isArray` (I've renamed it to `inArray` as that makes more sense to me, please let me know if you want me to revert that) a private field of the class which you pass in the constructor. We also don't want the user to do that so I renamed the class and removed its export, what we are exporting then is a `Validator` class which wraps the old class but it removes the `inArray` parameter so that the user doesn't see that at all (since it is an implementation details I think).

What do you think? :slightly_smiling_face:

dario-piotrowicz replied:
Actually... I don't think that the `Validator` class is accessible from user code is it? in that case I think that the extra class is not really necessary...

dario-piotrowicz replied:
extra class removed

yusukebe replied:
@dario-piotrowicz 

Renaming `isArray` to `inArray` and changing its accessibility of it are great! The `Validator` class must be accessed from `middleware/validator/middleware.ts`, so I think it's OK to keep it.

I want to go with this current code, OK?

dario-piotrowicz replied:
Yes, if you're ok with it I think it should be good to get merged :smiley:

yusukebe replied:
OK!

In addition, the most important reason to export `Validator` is there are use cases that the developer can write validation schema outside of the `validator` function.

```ts
const schema = (v: Validator) => ({
  title: v.body('title').isRequired(),
  body: v.body('body').isRequired(),
})

app.post('/posts', validator(schema), (c) => {
  const { title, body } = c.req.valid()
  //...
})
```

dario-piotrowicz replied:
ah I see, thanks for clarifying :slightly_smiling_face::+1:
